File: problems/golang/11.binary-tree/main/find_max_depth.go
package main

import (
	"math"

	"github.com/AmirAhmadzadeh/problems/binary_tree"
	"github.com/AmirAhmadzadeh/problems/node"
)

func calculateMaxDepth(node *node.Node, max float64) float64 {
	if node == nil {
		return max
	}

	max += 1
	return math.Max(calculateMaxDepth(node.Right(), max), calculateMaxDepth(node.Left(), max))
}

func findMaxDepth(tree *binary_tree.BinaryTree) int {
	return int(calculateMaxDepth(tree.GetRoot(), 0.0))
}

File: problems/ts/11.binary-tree/findMaxDepth.ts
import { Node } from './';

export function findMaxDepth(node: Node<number> | null, max = 0): number {
    if (!node) return max;

    max = max + 1;

    return Math.max.apply(null, [
        findMaxDepth(node.left, max),
        findMaxDepth(node.right, max),
    ]);
}
